packages/ui-concerto/src/lib/components/concertoForm.js

```diff
@@ -13,213 +13,156 @@
+    setValue(valueClone);
+    onValueChange(valueClone);
+  }, [onValueChange, value]);
+
+  // Default values which can be overridden by parent components
+  const options = React.useMemo(() => ({
+    includeOptionalFields: true,
+    includeSampleData: 'sample',
+    disabled: props.readOnly,
+    visitor: new ReactFormVisitor(),
+    onFieldValueChange: (e, key) => {
+      onFieldValueChange(e, key);
+    },
+    addElement: (e, key, field) => {
+      addElement(e, key, field);
+    },
+    removeElement: (e, key, index) => {
+      removeElement(e, key, index);
+    },
+    ...props.options,
+  }), [addElement, onFieldValueChange, removeElement, props.options, props.readOnly]);
+
+  const generator = React.useMemo(() => {
+    if (modelManager) {
+      return new FormGenerator(modelManager, options);
+    }
+    return null;
+  }, [modelManager, options]);
+
+  useEffect(() => {
+    setLoading(true);
+    const modelManager = new ModelManager();
+    // TODO Refactor this to an option to make this independent of Cicero
+    modelManager.addModelFile(
+      `namespace org.accordproject.base
+    abstract asset Asset {  }
+    abstract participant Participant {  }
+    abstract transaction Transaction identified by transactionId {
+      o String transactionId
+    }
+    abstract event Event identified by eventId {
+      o String eventId
+    }`,
+      'org.accordproject.base.cto',
+      false,
+      true
+    );
+    props.models.forEach((model, idx) => {
+      const decodedModel = decodeHTMLEntities(model);
+      modelManager.addModelFile(decodedModel, `model-${idx}`, true);
+    });
+    modelManager.updateExternalModels().then(() => {
+      setModelManager(modelManager);
+      setLoading(false);
+    });
+  }, [props.models]);
 
-  generateJSON(type) {
-    try {
-      // The type changed so we have to generate a new instance
-      if (this.props.json && !this.isInstanceOf(this.props.json, type)) {
-        return this.generator.generateJSON(type);
+  useEffect(() => {
+    if (props.type && generator) {
+      if (!value) {
+        const newJSON = generator.generateJSON(props.type);
+        setValue(newJSON);
       }
-      // The instance is null so we have to create a new instance
-      if (!this.props.json) {
-        return this.generator.generateJSON(type);
-      }
-    } catch (err) {
-      console.error(err);
     }
-    // Otherwise, just use what we already have
-    return this.props.json;
-  }
+  }, [generator, props.type, value]);
 
-  render() {
-    if (this.state.loading) {
-      return (
+  if (loading) {
+    return (
         <Dimmer active inverted>
           <Loader inverted>Loading</Loader>
         </Dimmer>
-      );
-    }
+    );
+  }
 
-    if (this.props.type && this.state.value) {
-      try {
-        return (
-          <Form style={{ minHeight: '100px', ...this.props.style }}>
-            {this.generator.generateHTML(this.props.type, this.state.value)}
+  if (props.type && value) {
```

Comment:
Can `props.type` ever be null?

Reply:
I added a new warning message if there is no `props.type`.